### PR TITLE
BAAS-23396: add strings overhead to mem usage calculations

### PR DESCRIPTION
**Pull request by Gabri3l**
This is a small ma pretty important change. Strings in go have an overhead of 16bytes, currently an empty string adds up to 0 bytes from the way we calculate memory usage which is not correct. With this change every native string or object key (which are also strings) will also account for the overhead defined in `SizeString`.

**EDIT**

To understand the impact of this change all `MemUsage` functions return the existing value + an adjusted one. This will help us understand the impact of such a change before we roll it out to customers. I split the commit logically for easier review but beware that the whole test suite only passes with the last commits.

* Original commit that has been already reviewed
* Update the interfaces to the new MemUsage signature
* Update every single MemUsage function to use the new signature + add tests
* A small refactor based on Nick's comment

Generally I noticed a few missing `nil` checks and other incorrect MemUsage functions, the diff should be easy to spot from the single commits.

### CODE REVIEW

File: memory_test.go [hunk 1]
```diff
@@ -613,64 +645,62 @@ func TestMemObjectsWithPropsLenThreshold(t *testing.T) {
...
 			200,
 			memUsageLimit,
-			// len("i100") + value i
-			50*4 + 50*SizeNumber +
-				// len("i150") + len("i150") + 150 number + len("i")
-				50*4 + 50*4 + 3 + 1 +
-				// for loop 100 + 200 + len("i")
-				SizeNumber + SizeNumber + 1,
+			// object overhead + key len("i100") + value i for the items 100 to 149
+			SizeEmpty + 50*(4+SizeString) + 50*SizeNumber +
```
Comment (LijieZhang1998) [inline memory_test.go:699]: [q] The question is not related to your change. But do you know why the multiplier is `50`? Thanks.

**Review by LijieZhang1998 — APPROVED** [review]
What a wonderful find! I only have one question but not really related to your change.

File: object.go [hunk 2]
```diff
@@ -1944,7 +1944,7 @@ func (o *baseObject) estimateMemUsage(ctx *MemUsageContext) (uint64, error) {
 		inc, err := v.MemUsage(ctx)
 		samplesVisited += 1
 		total += inc
-		total += uint64(len(k))
+		total += uint64(len(k)) + SizeString
 		averageMemUsage = float32(total) / float32(samplesVisited)
```
Comment (nickpoindexter) [inline object.go:1948]: [opt] I think you can avoid doing this calculation on every loop and only do it on the return. Sample function to keep it consistent.
```go
func roughMemUsage(sampleTotal, numSamples uint64, totalProps int) uint64 {
  return uint64(math.Ceil((float32(sampleTotal) / float32(numSamples)) * float32(totalProps)))
}
```

File: memory_test.go (hunk 1 above)
Comment (nickpoindexter) [reply memory_test.go:699]: for the first 50 entries the value is just `i` but after that it is `"i"+i`

**Review by nickpoindexter — APPROVED** [review]

**Review by jwongmongodb — APPROVED** [review]
LGTM! nothing else to add on my side. Great find!

**Review by kpatel71716 — APPROVED** [review]
LGTM! 🤑

File: array.go [hunk 3]
```diff
@@ -573,70 +573,78 @@ func (a *arrayObject) estimateMemUsage(ctx *MemUsageContext) (uint64, error) {
...
-		runningEstimate = uint64((float32(total) / float32(samplesVisited)) * float32(len(a.values)))
+		runningEstimate = uint64((float32(memUsage) / float32(samplesVisited)) * float32(len(a.values)))
+		newRunningEstimate = uint64((float32(newMemUsage) / float32(samplesVisited)) * float32(len(a.values)))
 		if err != nil {
-			return runningEstimate, err
+			return runningEstimate, newRunningEstimate, err
 		}
-		if exceeded := ctx.MemUsageLimitExceeded(total); exceeded {
-			return total, nil
+		if exceeded := ctx.MemUsageLimitExceeded(memUsage); exceeded {
```
Comment (LijieZhang1998) [inline array.go:586]: [q] should we use `mewMemUsage` instead of `memUsage`?

**Review by LijieZhang1998 — APPROVED** [review]
LGTM 🥇 I just have one question about the `MemUsageLimitExceeded` check.

File: array_test.go [hunk 4]
```diff
@@ -133,38 +133,63 @@ func BenchmarkArraySetEmpty(b *testing.B) {
...
-			mu:   NewMemUsageContext(New(), 88, 5000, 50, 50, TestNativeMemUsageChecker{}),
+			mu:   NewMemUsageContext(vm, 88, 5000, 50, 50, TestNativeMemUsageChecker{}),
 			ao: &arrayObject{
 				values: []Value{
-					New()._newString(newStringValue("key"), nil),
+					vm._newString(newStringValue("key"), nil),
 				},
 			},
-			expected:    41,
+			// array overhead + array baseObject
```
Comment (nickpoindexter) [inline array_test.go:154]: [q] How did we determine the array overhead is 8 bytes when I run it I get 24 bytes?

File: array.go (hunk 3 above)
Comment (Gabri3l) [reply array.go:586]: Good question, eventually the newMem and mem will have the same value but we are using this PR to collect data about the gap between the two values for customers to get an idea of the impact of this change. Until we apply this new changes all the memory logic will rely on the current memUsage.

File: string_test.go [hunk 5]
```diff
@@ -164,3 +164,42 @@ func BenchmarkASCIIConcat(b *testing.B) {
...
+	vm := New()
+
+	for _, tc := range []struct {
+		name           string
+		val            *stringObject
+		expectedMem    uint64
+		expectedNewMem uint64
+	}{
+		{
+			"should return SizeEmpty given a nil stringObject",
```
Comment (jwongmongodb) [inline string_test.go:178]: [opt] I find it easier to have named fields for readability

File: date.go [hunk 6]
```diff
@@ -174,15 +174,16 @@ func (d *dateObject) timeUTC() time.Time {
...
 	// start with the size of msec
-	total := SizeNumber
-	inc, err := d.baseObject.MemUsage(ctx)
-	total += inc
-	return total, err
+	memUsage = SizeNumber
+	newMemUsage = SizeNumber
+	inc, newInc, err := d.baseObject.MemUsage(ctx)
+
+	return memUsage + inc, newMemUsage + newInc, err
```
Comment (nickpoindexter) [inline date.go:188]: [nit] Can we add a test to check these return changes?

File: mem_context.go [hunk 7]
```diff
@@ -55,38 +55,6 @@ type NativeMemUsageChecker interface {
 	NativeMemUsage(goNativeValue interface{}) (uint64, bool)
 }
 
-func (self *stash) MemUsage(ctx *MemUsageContext) (uint64, error) {
```
Comment (nickpoindexter) [inline mem_context.go:58]: [q] Was this never used or it's replaced somewhere in this PR?

File: string_ascii_test.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+func TestStringAsciiMemUsage(t *testing.T) {
+	tests := []struct {
+		name        string
+		val         asciiString
+		expected    uint64
+		newExpected uint64
+		errExpected error
+	}{
+		{
+			name:        "should have a value given the length of the string",
```
Comment (nickpoindexter) [inline string_ascii_test.go:16]: [opt] Add test for empty string.

File: string_imported_test.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+	tests := []struct {
+		name        string
+		val         *importedString
+		expected    uint64
+		newExpected uint64
+		errExpected error
+	}{
+		{
+			name:        "should have a value given the length of the string",
+			val:         &importedString{s: "yo"},
```
Comment (nickpoindexter) [inline string_imported_test.go:17]: [opt] add test for empty string

File: string_unicode_test.go [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+		name        string
+		val         unicodeString
+		expected    uint64
+		newExpected uint64
+		errExpected error
+	}{
+		{
+			name:        "should have a value given the length of the string",
+			val:         unicodeString{' ', 'y', 'o'},
+			expected:    2,
```
Comment (nickpoindexter) [inline string_unicode_test.go:18]: [q] why 2 instead of 3? Does space not count?

**Review by nickpoindexter — APPROVED** [review]
LGTM! Just had some questions and optional things.

File: runtime.go [hunk 11]
```diff
@@ -512,42 +512,52 @@ func (r *Runtime) typeErrorResult(throw bool, args ...interface{}) {
...
+			return memUsage, newMemUsage, err
 		}
 	}
 
-	for idx := range r.vm.callStack {
-		inc, err := r.vm.callStack[idx].MemUsage(ctx)
-		total += inc
-		if err != nil {
-			return total, err
+	if r.vm != nil {
```
Comment (jwongmongodb) [inline runtime.go:529]: [opt] can do an early exit to remove the additional layer of conditional blocks.

```
if r.vm == nil{
  return memUsage, newMemUsage, nil
}
```

File: mem_context.go (hunk 7 above)
Comment (nickpoindexter) [reply mem_context.go:58]: Never mind, I see it was just moved.

File: builtin_map.go [hunk 12]
```diff
@@ -94,19 +94,19 @@ func (mo *mapObject) exportToMap(dst reflect.Value, typ reflect.Type, ctx *objec
...
 	if err := ctx.Descend(); err != nil {
-		return 0, err
+		return 0, 0, err
 	}
 
-	total, err := mo.baseObject.MemUsage(ctx)
+	memUsage, newMemUsage, err = mo.baseObject.MemUsage(ctx)
 	if err != nil {
-		return total, err
+		return memUsage, newMemUsage, err
```
[the remark this answers is not in the document]
Comment (Gabri3l) [reply builtin_map.go:109]: No reason, just a miss, nice!

File: builtin_map.go [hunk 13]
```diff
@@ -94,19 +94,19 @@ func (mo *mapObject) exportToMap(dst reflect.Value, typ reflect.Type, ctx *objec
...
+func (mo *mapObject) MemUsage(ctx *MemUsageContext) (memUsage uint64, newMemUsage uint64, err error) {
 	if mo == nil || ctx.IsObjVisited(mo) {
-		return SizeEmpty, nil
+		return SizeEmpty, SizeEmpty, nil
 	}
 	ctx.VisitObj(mo)
 
 	if err := ctx.Descend(); err != nil {
-		return 0, err
+		return 0, 0, err
```
[the remark this answers is not in the document]
Comment (Gabri3l) [reply builtin_map.go:104]: You gave me an idea, I might rename `SizeEmpty` beacuse it just refers to an empty `baseObject` and its value is not 0 but 8.

Maybe `SizeEmptyBaseObj` would make more sense?

[the remark this answers is not in the document]
Comment (nickpoindexter) [reply builtin_map.go:104]: `SizeEmptyObj`, `SizeEmptyObject`, or `SizeEmptyStruct` all work for me 😄

File: array_test.go (hunk 4 above)
Comment (nickpoindexter) [reply array_test.go:154]: I just empty arrays of different [types in the playground](https://go.dev/play/p/BZE3v0A71g2). I wasn't looking at that overall expected. just the `array overhead` part where that value seems to be 8.

File: date.go (hunk 6 above)
Comment (Gabri3l) [reply date.go:188]: Do you mean a case with a non-empty baseObject (which I noticed was missing)?

File: string_unicode_test.go (hunk 10 above)
Comment (Gabri3l) [reply string_unicode_test.go:18]: When this gets converted to a string (see the String() method on this type) we skip the first item which can represent the little-indian/big-indian convention but not the actual content of the string. Not 100% that's what the first char is used for but I know that even comparisons between strings skip the fist char if it's a unicode string.

File: date.go (hunk 6 above)
Comment (Gabri3l) [reply date.go:188]: perfect, will add a test!

File: array_test.go (hunk 4 above)
Comment (Gabri3l) [reply array_test.go:154]: Ah ok! So the `arrayObject` is a struct and could potentially have `values` be `nil`. But if it is not `nil` then you're right there's an overhead of just storing an empty array of 24 bytes! I'll take another look at this and keep you posted.

File: builtin_map.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (Gabri3l) [reply builtin_map.go:104]: `baseObject` and `Object` are 2 different types so I think at this point I prefer `SizeEmptyStruct`

File: array.go [hunk 14]
```diff
@@ -573,70 +573,78 @@ func (a *arrayObject) estimateMemUsage(ctx *MemUsageContext) (uint64, error) {
...
+			return memUsage, newMemUsage, nil
 		}
 	}
 
-	return runningEstimate, nil
+	return runningEstimate, newRunningEstimate, nil
```
[the remark this answers is not in the document]
Comment (Gabri3l) [reply array.go:591]: Sorry not sure I follow, if there's an error when calculating the mem usage of a sample we need to know the estimate at that point in time so we can exit early and still provide an estimate given what we know.

Also made me realize of a change I need to make in estimate functions, sweet!

[the remark this answers is not in the document]
Comment (Gabri3l) [reply array.go:591]: Resolved offline!

**Review by nickpoindexter — APPROVED** [review]
LGTM! Will be interesting to see the results for customers

File: memory_test.go [hunk 15]
```diff
@@ -287,164 +348,205 @@ func TestMemCheck(t *testing.T) {
...
 		{
-			"Native value",
-			`checkMem();
+			desc: "Native value",
+			script: `checkMem();
 			nv = new MyNativeVal()
 			checkMem();`,
-			testNativeValueMemUsage + 2,
+			expectedSizeDiff: testNativeValueMemUsage +
+				2, // "nv"
```
Comment (jwongmongodb) [inline memory_test.go:468]: [q] what does `nv` mean here? Native Value?

**Review by jwongmongodb — APPROVED** [review]
LGTM!

**Review by kpatel71716 — APPROVED** [review]
LGTM!

File: memory_test.go (hunk 15 above)
Comment (Gabri3l) [reply memory_test.go:468]: this is for the variable name `nv`, so it's the len("nv")
